### PR TITLE
[REEF-1859] Support Avro code generation of multiple C# files from si…

### PR DESCRIPTION
**Pull request by dougmsft**
…ngle avsc.

  * This task modifies the code generation task to support generating
    multiple C# program files from a single avsc file that has multiple
    parallel and nested types.
  * Adds the bridge message wrapper avsc file that code gens to four C# files
    and provides the framework for the bridge Java/CLR Avro protocol.

JIRA:
  [REEF-1859](https://issues.apache.org/jira/browse/REEF-1859)

Pull request:
  This closes #

### CODE REVIEW

File: lang/common/bridge/avro/BridgeInterop.avsc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  /*
+   * Identify the next message in the Java/C# bridge protocol.
+   */
+  { "type": "record", "name":"BridgeInterop", "namespace": "org.apache.reef.bridge.message",
+    "doc": "The BridgeInterop message is the container for all Java/CLR interop messages.",
+    "fields": [
+      { "name": "messageType", "type":
+        { "type": "enum", "name": "MessageType", "symbols": ["SetupBridge","SystemOnStart"]},
+        "doc": "The type of the bridge Java/CLR interop message contained in the wrapper message." },
+      { "name": "setupBridge", "type": { "type": "record", "name": "SetupBridge",
```
Comment (markusweimer) [inline lang/common/bridge/avro/BridgeInterop.avsc]: Can these types be defined in their own right and then just referenced here?

File: lang/cs/AvroCodeGeneration.targets [hunk 2]
```diff
@@ -82,41 +82,57 @@ under the License.
         bool conversionError = false;
         using (StreamWriter logWriter = new StreamWriter(Path.Combine(tempDir, "avroout.log")))
         {
-          logWriter.WriteLine(Environment.GetEnvironmentVariable("path"));
-          foreach (String file in sourceFiles)
+          //logWriter.WriteLine(Environment.GetEnvironmentVariable("path"));
```
Comment (markusweimer) [inline lang/cs/AvroCodeGeneration.targets]: please don't comment out code. Either delete it or keep it.

File: lang/cs/AvroCodeGeneration.targets [hunk 3]
```diff
@@ -82,41 +82,57 @@ under the License.
...
-              logWriter.WriteLine(stdErrBuilder.ToString());
+                StringBuilder stdOutBuilder = new StringBuilder();
+                StringBuilder stdErrBuilder = new StringBuilder();
 
-              using (StreamReader tmpReader = new StreamReader(Path.Combine(tempDir, Path.GetFileName(file))))
-              {
-                String source = tmpReader.ReadToEnd();
-                using (StreamWriter destWriter = new StreamWriter(Path.GetFullPath(file)))
+                // Start the conversion process
+                using (Process rProc = Process.Start(rProcInfo))
```
Comment (markusweimer) [inline lang/cs/AvroCodeGeneration.targets]: Any chance we can start these processes without opening command windows?

File: lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs [hunk 4]
```diff
@@ -28,7 +28,7 @@ namespace org.apache.reef.bridge.message
     [DataContract(Namespace = "org.apache.reef.bridge.message")]
     public partial class SystemOnStart
     {
-        private const string JsonSchema = @"{""type"":""record"",""name"":""org.apache.reef.bridge.message.SystemOnStart"",""doc"":""The SystemOnStart message is sent to the CLR bridge from the Java bridge when the round trip communication path has been established."",""fields"":[{""name"":""dateTime"",""doc"":""Date time in seconds as a long since January 1, 1970"",""type"":""long""}]}";
+        private const string JsonSchema = @"{""type"":""record"",""name"":""org.apache.reef.bridge.message.SystemOnStart"",""fields"":[{""name"":""dateTime"",""doc"":""Date time in seconds as a long since January 1, 1970"",""type"":""long""}]}";
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs]: Why did we loose the doc string on it?

File: lang/common/bridge/avro/BridgeInterop.avsc (hunk 1 above)
Comment (dougmsft) [inline lang/common/bridge/avro/BridgeInterop.avsc]: I could not find a way to put them in a different file but I think I can define them in the same file before the BridgeInterop record so I will give that a try.

File: lang/cs/AvroCodeGeneration.targets (hunk 2 above)
Comment (dougmsft) [inline lang/cs/AvroCodeGeneration.targets]: Removed

File: lang/cs/AvroCodeGeneration.targets (hunk 3 above)
Comment (dougmsft) [inline lang/cs/AvroCodeGeneration.targets]: I will investigate. I believe there is a way to do that.

File: lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs (hunk 4 above)
Comment (dougmsft) [inline lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs]: I am not sure and will investigate.

File: lang/common/bridge/avro/BridgeInterop.avsc (hunk 1 above)
Comment (dougmsft) [inline lang/common/bridge/avro/BridgeInterop.avsc]: I have separated them out in the same file.

File: lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs (hunk 4 above)
Comment (dougmsft) [inline lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs]: The comments end up in the java generated sources. Looks like the Microsoft version of Avro does not pass them through.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Bridge.CLR/Message/SystemOnStart.cs]: Bummer. We should file a bug about that :)

File: lang/common/bridge/avro/BridgeInterop.avsc (hunk 1 above)
Comment (dougmsft) [inline lang/common/bridge/avro/BridgeInterop.avsc]: For some reason, the MessageType enumeration has to be in line for the Java code generation to work, but all of the records have been moved to the head of the file. I also put the messages into a union.

File: lang/cs/AvroCodeGeneration.targets (hunk 3 above)
Comment (dougmsft) [inline lang/cs/AvroCodeGeneration.targets]: `CreateNoWindow = true` property on the ProcessStartInfo class does the trick.

Comment (markusweimer) [inline lang/cs/AvroCodeGeneration.targets]: Very nice!

**Comment by dougmsft** [comment]
This one is good to go.

**Pull request closed without merging**